include/dlaf/sender/transform.h

```diff
@@ -25,210 +26,81 @@
-    template <typename... Ts>
-    static constexpr bool is_cuda_stream_invocable =
-        std::is_invocable_v<unwrapping_function_type, std::decay_t<Ts>&..., cudaStream_t>;
-    template <typename... Ts>
-    static constexpr bool is_cublas_handle_invocable =
-        std::is_invocable_v<unwrapping_function_type, cublasHandle_t, std::decay_t<Ts>&...>;
-    template <typename... Ts>
-    static constexpr bool is_cusolver_handle_invocable =
-        std::is_invocable_v<unwrapping_function_type, cusolverDnHandle_t, std::decay_t<Ts>&...>;
-    template <typename... Ts>
-    static constexpr bool is_gpu_invocable =
-        is_cuda_stream_invocable<Ts...> || is_cublas_handle_invocable<Ts...> ||
-        is_cusolver_handle_invocable<Ts...>;
+/// Lazy transform. This does not submit the work and returns a sender.
+template <Backend B, typename F, typename Sender,
+          typename = std::enable_if_t<pika::execution::experimental::is_sender_v<Sender>>>
+[[nodiscard]] decltype(auto) transform(const Policy<B> policy, F&& f, Sender&& sender) {
+  using pika::execution::experimental::then;
+  using pika::execution::experimental::transfer;
+  using pika::unwrapping;
 
-    template <typename G, typename... Us>
-    static auto call_helper(cudaStream_t stream, cublasHandle_t cublas_handle,
-                            cusolverDnHandle_t cusolver_handle, G&& g, Us&... us) {
-      static_assert(is_gpu_invocable<Us...>,
-                    "function passed to transform<GPU> must be invocable with a cudaStream_t as the "
-                    "last argument or a cublasHandle_t/cusolverDnHandle_t as the first argument");
+  auto scheduler = getBackendScheduler<B>(policy.priority());
+  auto transfer_sender = transfer(std::forward<Sender>(sender), std::move(scheduler));
+  auto f_unwrapping = pika::unwrapping(std::forward<F>(f));
 
-      if constexpr (is_cuda_stream_invocable<Us...>) {
-        (void) cublas_handle;
-        (void) cusolver_handle;
-        return std::invoke(pika::unwrapping(std::forward<G>(g)), us..., stream);
-      }
-      else if constexpr (is_cublas_handle_invocable<Us...>) {
-        (void) cusolver_handle;
-        (void) stream;
-        return std::invoke(pika::unwrapping(std::forward<G>(g)), cublas_handle, us...);
-      }
-      else if constexpr (is_cusolver_handle_invocable<Us...>) {
-        (void) cublas_handle;
-        (void) stream;
-        return std::invoke(pika::unwrapping(std::forward<G>(g)), cusolver_handle, us...);
-      }
+  if constexpr (B == Backend::MC) {
+    return then(std::move(transfer_sender), std::move(f_unwrapping));
+  }
+  else if constexpr (B == Backend::GPU) {
+#if defined(DLAF_WITH_CUDA)
+    using pika::cuda::experimental::then_with_cublas;
+    using pika::cuda::experimental::then_with_cusolver;
+    using pika::cuda::experimental::then_with_stream;
+
+    if constexpr (std::is_invocable_v<decltype(then_with_stream), decltype(std::move(transfer_sender)),
```

Comment:
Note to @auriane (and FYI for others), this used `then_with_any_cuda` for a moment but because of the problems with some of the HIP/ROCm handles being `void*` and this needing explicit dispatching I've reverted the choice of CUDA stream/cuBLAS handle/cuSOLVER handle back to DLA-Future. This will be something like `if constexpr (dispatch_type == cuda_stream)` after #459.